File: test_dir/cpp/src/loops/reversed_.cpp
#include "loops\reversed_.h"
#include "py_list.h"
#include "py_reversed.h"
#include "py_str.h"
#include "py_zip.h"
#include "pypp_util/print.h"

void reversed_fn() {
    print(PyStr("REVERSED RESULTS:"));
    PyList<int> a = PyList<int>({});
    for (const auto &x : PyReversed(PyList({1, 2, 3}))) {
        a.append(x);
    }
    print(a);
    PyList<PyStr> b = PyList<PyStr>({});
    for (const auto &x : PyReversed(PyStr("abcd"))) {
        b.append(x);
    }
    print(b);
    for (const auto &[x, z] :
         PyZip(PyReversed(PyList({1, 2, 3})), PyReversed(PyStr("abc")))) {
        print(PyStr(std::format("{}, {}", x, z)));
    }
}

File: test_dir/python/src/loops/reversed_.py
from test_dir.python.pypp.ownership import mov


def reversed_fn():
    print("REVERSED RESULTS:")
    # over a list
    a: list[int] = []
    for x in reversed([1, 2, 3]):
        y: int = x
        a.append(mov(y))
    print(a)
    # over a string
    b: list[str] = []
    for x in reversed("abcd"):
        y: str = x
        b.append(mov(y))
    print(b)
    # with zip
    for x, z in zip(reversed([1, 2, 3]), reversed("abc")):
        print(f"{x}, {z}")
